forms-flow-api/src/formsflow_api/services/keycloak_service.py

```diff
@@ -0,0 +1,97 @@
+"""Keycloak connect and operations services for import data."""
+from http import HTTPStatus
+from marshmallow.exceptions import ValidationError
+from flask import current_app
+from formsflow_api.services.factory import KeycloakFactory
+from formsflow_api.schemas import (
+    KeycloakResponseSchema, KeycloakUserSchema,
+    RequestUserSchema
+)
+
+class KeycloakService:  # pylint: disable=too-few-public-methods
+    """This class manages keycloak service."""
+
+    def __init__(self):
+        """Initialize client."""
+        self.client = KeycloakFactory.get_instance()
+    
+    def get_group_id(self, name):
+        """fetch user from group name"""
+        try:
+            group = self.client.get_group_by_name(name)
+            return group.get('id')
+        except:
+            return None
+    
+    def get_user_id(self, username):
+        """fetch user from username, exact match"""
+        users = self.client.get_user_by_username(
+            username = username
+        )
+        userId = None
+        if len(users) > 0:
+            userId = users[0]["id"]
+        return userId
+
+    def get_user_or_add(self, userData):
+        """get user and if not exist and add new one"""
+        result = KeycloakResponseSchema().load({})
+        try:
+            userData = RequestUserSchema().load(userData)
+        except ValidationError as e:
+            fields = ",".join(e.messages.keys())
+            result["message"] = f"{fields} fields required with valid data"
+            return result
+
+        username = userData.get("email")
+        idp = userData.get("idp")
+        if idp and idp !="":
+            username = f"{username}_{idp}"
```

Comment:
This only works for `IDIR` users. for non-idir users at this point, Keycloak only merges based on the username and we don't know their username until they log in.

A workaround would be to find out how to make Keycloak merge accounts based on the email in each IDP.

Reply:
This is fine for now because for the Senior Leader form, we only need IDIRs. We can improve this later.

Reply:
Please limit this feature for IDIR at the moment until we figure out how to scale it for other IDPs.

Reply:
limited this feature for IDIR only